applications/sdp/mspi/src/main.c

```diff
@@ -52,215 +57,494 @@ typedef struct __packed {
+	case MSPI_IO_MODE_HEX:
+	case MSPI_IO_MODE_HEX_8_8_16:
+	case MSPI_IO_MODE_HEX_8_16_16:
+	case MSPI_IO_MODE_MAX:
+	default:
+		/* TODO: Jira ticket: NRFX-6875 error*/
+		break;
+	}
 
-static void process_packet(const void *data, size_t len, void *priv)
+	return xfer_mode;
+}
+
+static void ep_bound(void *priv)
+{
+	atomic_set_bit(&ipc_atomic_sem, NRFE_MSPI_EP_BOUNDED);
+}
+
+static void ep_recv(const void *data, size_t len, void *priv)
 {
 	(void)priv;
-	(void)len;
-	nrfe_mspi_flpr_response_t response;
-	uint8_t opcode = *(uint8_t *)data;
 
-	response.opcode = opcode;
+	process_packet(data, len);
+}
 
-	switch (opcode) {
-	case NRFE_MSPI_CONFIG_PINS: {
-		/* TODO: Process pinctrl config data
-		 * nrfe_mspi_pinctrl_soc_pin_t *pins_cfg = (nrfe_mspi_pinctrl_soc_pin_t *)data;
-		 * response.opcode = pins_cfg->opcode;
-		 *
-		 * for (uint8_t i = 0; i < NRFE_MSPI_PINS_MAX; i++) {
-		 *         uint32_t psel = NRF_GET_PIN(pins_cfg->pin[i]);
-		 *         uint32_t fun = NRF_GET_FUN(pins_cfg->pin[i]);
-		 *         NRF_GPIO_Type *reg = nrf_gpio_pin_port_decode(&psel);
-		 * }
-		 */
+static void ll_prepare_transfer(enum base_io_mode xfer_mode, uint32_t data_length)
+{
+	uint16_t dir;
+	uint16_t out;
+	nrf_vpr_csr_vio_config_t config;
+	nrf_vpr_csr_vio_mode_out_t out_mode = {
+		.mode = NRF_VPR_CSR_VIO_SHIFT_OUTB_TOGGLE,
+		.frame_width = 1,
+	};
+
+	dir = nrf_vpr_csr_vio_dir_get();
+	out = nrf_vpr_csr_vio_out_get();
+
+	switch (xfer_mode) {
+	case BASE_IO_MODE_SINGLE:
+		out_mode.frame_width = 1;
 		break;
-	}
-	case NRFE_MSPI_CONFIG_CTRL: {
-		/* TODO: Process controller config data
-		 * nrfe_mspi_cfg_t *cfg = (nrfe_mspi_cfg_t *)data;
-		 * response.opcode = cfg->opcode;
-		 */
+	case BASE_IO_MODE_DUAL:
+		out_mode.frame_width = 2;
 		break;
-	}
-	case NRFE_MSPI_CONFIG_DEV: {
-		/* TODO: Process device config data
-		 * nrfe_mspi_dev_cfg_t *cfg = (nrfe_mspi_dev_cfg_t *)data;
-		 * response.opcode = cfg->opcode;
-		 */
+	case BASE_IO_MODE_QUAD:
+		out_mode.frame_width = 4;
 		break;
-	}
-	case NRFE_MSPI_CONFIG_XFER: {
-		/* TODO: Process xfer config data
-		 * nrfe_mspi_xfer_t *xfer = (nrfe_mspi_xfer_t *)data;
-		 * response.opcode = xfer->opcode;
-		 */
+	case BASE_IO_MODE_OCTAL:
+		out_mode.frame_width = 8;
 		break;
 	}
-	case NRFE_MSPI_TX:
-	case NRFE_MSPI_TXRX: {
-		nrfe_mspi_xfer_packet_t *packet = (nrfe_mspi_xfer_packet_t *)data;
 
-		response.opcode = packet->opcode;
+	NRFX_ASSERT(data_vios_count >= out_mode.frame_width);
+	NRFX_ASSERT(data_length % out_mode.frame_width == 0)
 
-		if (packet->packet.dir == MSPI_RX) {
-			/* TODO: Process received data */
-		} else if (packet->packet.dir == MSPI_TX) {
-			/* TODO: Send data */
-		}
-		break;
+	uint16_t direction = (nrfe_mspi_xfer_packet.dir == MSPI_TX) ? VPRCSR_NORDIC_DIR_OUTPUT
+								    : VPRCSR_NORDIC_DIR_INPUT;
+
+	for (uint8_t i = 0; i < out_mode.frame_width; i++) {
+		dir = BIT_SET_VALUE(dir, data_vios[i], direction);
+		out = BIT_SET_VALUE(out, data_vios[i], VPRCSR_NORDIC_OUT_LOW);
```

Comment:
IMO, It is not good approach to set all pins to input or output depending on transfer type.

Reply:
what do you mean? in tx transfer all data pins are output, in RX transfer all data pins are input

Reply:
I agree with @jaz1-nordic. Peripheral device should not transfer data to FLPR by itself, FLPR should first send some command or address. So RX actually consists of TX and RX and pins should not be as set input in RX. They first should be set as output and then as input.